### ISSUE TITLE
correct way to handle "SQ ring full" state

### ISSUE THREAD

**Issue by romange** [axboe/liburing#396]
Hi, 
suppose `io_uring_get_sqe()` returned NULL .  My ring is running without IORING_SETUP_SQPOLL flag.
What are my options? 

1. Does `io_uring_enter` call with min_complete=1  ensure that SQ is offloaded? 
2. Can SQ get full just because I have not called io_uring_enter() quickly enough, thus did not allow kernel to offload SQ ring?

**Comment by axboe** [comment]
1) You can either wait for something to complete, or you can explicitly wait for SQ ring space with IORING_ENTER_SQ_WAIT
2) Right, the sq thread will only poll it's running. If it's gone to sleep it'll need a syscall to wake it up, and if that is the case, then it's quite possible to fill the SQ ring and get NULL returned on grabbing a new sqe. That's also possible in case the thread is running, as filling in new entries is faster than getting them submitted (more work is involved there, vs just writing out 64 bytes to memory).

Hope that helps!

**Comment by romange** [comment]
Thanks. based on https://man.archlinux.org/man/io_uring_enter.2.en it seems that IORING_ENTER_SQ_WAIT is applicable only for IORING_SETUP_SQPOLL rings.

**Comment by axboe** [comment]
Heh, I totally read your initial message as "with SQPOLL". What I wrote applies to SQPOLL being enabled. If you're not using SQPOLL and you get a NULL return when trying to get an sqe, then you already filled in all your entries. The only sane course of action there is to either increase the size of the sqring when you set it up, or just submit whatever you already have pending and then retry the sqe get. It'll succeed once you have the kernel consume entries from it.

**Comment by romange** [comment]
Thanks!

**Comment by YoSTEALTH** [comment]
Extra clarification,

Getting single `sqe` is simple enough you can do it something like:
https://github.com/YoSTEALTH/Liburing/blob/master/liburing/helper.py#L156

Multiple `sqes` can be tricky, this is how I went about it:
https://github.com/YoSTEALTH/Liburing/blob/master/liburing/helper.py#L118

**Comment by romange** [comment]
Thanks @YoSTEALTH  lets see if it's possible to improve the API so that it we won't need to do busy waits.

@axboe  I would like to get to bottom of this. Here is a specific scenario that I am failing to understand:

- I run io_uring *without* SQ_POLL of size 256.
- the flow just unloaded bunch of completions (64 of them) and called `io_uring_cq_advance(&ring_, cqe_count);`
- The sq ring is full, i.e. `io_uring_sq_ready(&ring_)` returns 256.
- I call `sys_io_uring_enter(ring_fd, 256, 0, 0, NULL);` but it returns -1 (EBUSY).

Questions:
1. Why after I advance kernel `cq->khead` and I context switch into kernel, it does not unload sq ring? I suspect that it's because unloading CQ does not immediately mean that the submit buffers become available. Just want to confirm it.
2. I appreciate the asynchronous nature of io_uring API, however, it seems that there are corner cases when blocking behavior can be helpful. For example, here I try to model periodic timer event. When it's fired, its callback function recharges it back using IORING_OP_TIMEOUT opcode. If I am not able to allocate an sqe from the callback function I can not continue with my flow. My basic axiom is: it is possible to allocate a single sqe from the completion callback because if completion callback runs it means I removed a single item from CQ buffer. So as long as I do not allocate more than a single entry it should work.  Maybe it make sense to use  `IORING_ENTER_SQ_WAIT` for non SQPOLL rings as well?
If not, how you would manage back pressure here?

**Comment by isilence** [comment]
@romange, it looks you get EBUSY, because there is a backlog, i.e. overflowed completions.
https://elixir.bootlin.com/linux/v5.12.19/source/fs/io_uring.c#L6667

So, you previously submitted so many requests that after they had completed, io_uring didn't find place in CQ to fit them, and so they got stashed internally. In this case it expects you to reap CQEs (including `io_uring_cq_advance`) and do `sys_io_uring_enter` again, so those overflowed completions got placed into CQ, and you can process them.

**Comment by isilence** [comment]
iow, you can do:

```C
ring = create_ring(sq_size=8, cq_size=8);
submit(8);
submit(8);
submit(8);
```

And submit __might__ not return any errors, requests will be inflight. Then, they will complete, 8 would be placed in CQ, and 16 CQEs stashed internally, so you would need to enter the kernel to get them.

And that check leading to -EBUSY was deleted in 5.13, so since 5.13 it won't return EBUSY in this scenario.

**Comment by romange** [comment]
Oh, that's good to know! Thanks, Pavel. I could rearrange my code and fix periodic timer implementation so it would be recharged after I unload all completions including stashed ones. However, in general, that check prevents CQE callbacks from submitting new requests, which is problematic, imho.

**Comment by isilence** [comment]
> * Why after I advance kernel `cq->khead` and I context switch into kernel, it does not unload sq ring? I suspect that it's because unloading CQ does not immediately mean that the submit buffers become available. Just want to confirm it.

Because, there are overflowed CQEs and the kernel expects you to reap them (via entering the kernel with waiting first), a kind of backpressure. You won't get the error for kernels 5.13+, because it sounded inconvenient indeed. But that change depends on some other features in 5.13, so I'm not sure can be backported to older kernels.

> * I appreciate the asynchronous nature of io_uring API, however, it seems that there are corner cases when blocking behavior can be helpful. For example, here I try to model periodic timer event. When it's fired, its callback function recharges it back using IORING_OP_TIMEOUT opcode. If I am not able to allocate an sqe from the callback function I can not continue with my flow. My basic axiom is: it is possible to allocate a single sqe from the completion callback because if completion callback runs it means I removed a single item from CQ buffer. So as long as I do not allocate more than a single entry it should work.  Maybe it make sense to use  `IORING_ENTER_SQ_WAIT` for non SQPOLL rings as well?

If we forget about EBUSY for now, `sys_io_uring_enter(submit=N)` tries hardest to submit those requests, so you wouldn't need to SQ_WAIT but just to submit requests. Thus I'd say it's not needed and the responsibility is on the userspace, especially considering that it already should be accessing SQ in a synchronised way (or single-threaded). 

I'd say there is no so direct link between number of CQEs and SQEs queued, but the remedy is again to submit if SQ is full.

**Comment by isilence** [comment]
> Oh, that's good to know! Thanks, Pavel. I could rearrange my code and fix periodic timer implementation so it would be recharged after I unload all completions including stashed ones. 

Great

> However, in general, that check prevents CQE callbacks from submitting new requests, which is problematic, imho.

Pretty much agree, it's much better to be able to freely submit (if there is place in SQ), and that's the behaviour since 5.13.

**Comment by isilence** [comment]
Was thinking whether we can backport it, but apart from simply taking
https://lore.kernel.org/io-uring/20210416232837.EpkCt1h-_Fk1PTRiQA92ifC9l5pNLc2vk84GqYJcJJc@z/

would need the one below to be taken, and doing so looks too error-prone for stable (considering differences in code bases).
https://lore.kernel.org/io-uring/20210317162943.173837-4-axboe@kernel.dk/

**Comment by romange** [comment]
It's fine really. We focus here on the edge case where ring buffers are not enough to handle the expected traffic. The workaround for older kernels exists  - to increase the ring size. I will still try to fix my lib to be more robust with < 5.13 kernels but if the layer above my lib will allocate SQEs in its completion callbacks it might crash the process when EBUSY is returned.

**Comment by romange** [comment]
Yes, thank you!
